Add support for unaligned/intrinsic to Deflate with RLE

nmoinvaz opened the pull request:
There is no AVX support because it was equal or slower to no intrinsics/unaligned-reads during my testing.

nmoinvaz commented:
Actually, I fixed some bugs and I am not seeing any performance benefit on a bitmap image.

nmoinvaz commented:
Rebased

nmoinvaz commented:
## RUN 1

### DEVELOP
```
 Tool: ../zlib-ng/build-develop/Release/minigzip.exe Levels: 0-0
 Runs: 70         Trim worst: 40

 Level   Comp   Comptime min/avg/max/stddev  Decomptime min/avg/max/stddev  Compressed size
 0    100.008%      0.221/0.244/0.255/0.010        0.222/0.256/0.269/0.010      211,973,953
 R     60.465%      3.097/3.715/4.071/0.282        1.151/1.227/1.252/0.024      128,159,347

 avg1  80.236%                        1.979                          0.742
 avg2 160.472%                        3.959                          1.484
 tot                                118.766                         44.508      340,133,300
```
### PR
```
 Tool: ../zlib-ng/build-rle/Release/minigzip.exe Levels: 0-0
 Runs: 70         Trim worst: 40

 Level   Comp   Comptime min/avg/max/stddev  Decomptime min/avg/max/stddev  Compressed size
 0    100.008%      0.240/0.255/0.268/0.009        0.244/0.262/0.272/0.008      211,973,953
 R     60.465%      2.934/3.326/3.571/0.169        1.157/1.215/1.249/0.025      128,159,347

 avg1  80.236%                        1.791                          0.738
 avg2 160.472%                        3.581                          1.477
 tot                                107.441                         44.299      340,133,300
```
Compression speed improvement 9.53%.

## RUN 2
### DEVELOP
```
 Tool: ../zlib-ng/build-develop/Release/minigzip.exe Levels: 0-0
 Runs: 70         Trim worst: 40

 Level   Comp   Comptime min/avg/max/stddev  Decomptime min/avg/max/stddev  Compressed size
 0    100.008%      0.142/0.147/0.150/0.002        0.140/0.149/0.156/0.005      211,973,953
 R     60.465%      1.920/2.060/2.251/0.124        0.788/0.800/0.807/0.004      128,159,347

 avg1  80.236%                        1.103                          0.475
 avg2 160.472%                        2.207                          0.949
 tot                                 66.207                         28.474      340,133,300
```
### PR
```
 Tool: ../zlib-ng/build-rle/Release/minigzip.exe Levels: 0-0
 Runs: 70         Trim worst: 40

 Level   Comp   Comptime min/avg/max/stddev  Decomptime min/avg/max/stddev  Compressed size
 0    100.008%      0.141/0.145/0.148/0.002        0.140/0.149/0.155/0.006      211,973,953
 R     60.465%      1.870/2.002/2.196/0.106        0.791/0.803/0.808/0.004      128,159,347

 avg1  80.236%                        1.074                          0.476
 avg2 160.472%                        2.148                          0.951
 tot                                 64.426                         28.542      340,133,300
```
Compression speed improvement of 2.69%.

nmoinvaz commented:
```
2022-03-17T19:14:09-07:00
Running C:\Users\nathan\Source\zlib-ng\build-rle\test\benchmarks\Release\benchmark_zlib.exe
Run on (8 X 2995 MHz CPU s)
CPU Caches:
  L1 Data 48 KiB (x4)
  L1 Instruction 32 KiB (x4)
  L2 Unified 1280 KiB (x4)
  L3 Unified 12288 KiB (x1)
--------------------------------------------------------------------------
Benchmark                                Time             CPU   Iterations
--------------------------------------------------------------------------
compare256_rle/c/1                    2.31 ns         2.02 ns    448000000
compare256_rle/c/8                    2.04 ns         1.96 ns    263529412
compare256_rle/c/64                   12.9 ns         12.3 ns     56000000
compare256_rle/c/256                  70.9 ns         69.8 ns      8960000
compare256_rle/unaligned_16/1         1.55 ns         1.53 ns    448000000
compare256_rle/unaligned_16/8         2.01 ns         1.99 ns    344615385
compare256_rle/unaligned_16/64        7.09 ns         6.98 ns     89600000
compare256_rle/unaligned_16/256       32.1 ns         31.5 ns     21333333
compare256_rle/unaligned_32/1         2.08 ns         2.09 ns    344615385
compare256_rle/unaligned_32/8         2.81 ns         2.83 ns    248888889
compare256_rle/unaligned_32/64        8.52 ns         8.58 ns     74666667
compare256_rle/unaligned_32/256       37.2 ns         37.7 ns     18666667
compare256_rle/unaligned_64/1         1.74 ns         1.72 ns    373333333
compare256_rle/unaligned_64/8         2.21 ns         2.20 ns    320000000
compare256_rle/unaligned_64/64        6.11 ns         6.00 ns    112000000
compare256_rle/unaligned_64/256       23.1 ns         23.4 ns     32000000

```

nmoinvaz committed:
Added unit tests for compare256_rle family of functions.

mtl1979 commented on compare256_rle.c:
```diff
@@ -0,0 +1,130 @@
+/* compare256_rle.c -- 256 byte rune length encoding comparison
```
Spelling: `run-length`

nmoinvaz replied:
Fixed. Thanks!

nmoinvaz commented:
After 1 year as a draft this is now ready for review. 🤣

Dead2 commented on deflate_rle.c:
```diff
@@ -34,21 +33,13 @@ Z_INTERNAL block_state deflate_rle(deflate_state *s, int flush) {
 
         /* See how many times the previous byte repeats */
-        if (s->lookahead >= STD_MIN_MATCH && s->strstart > 0) {
+        if (s->lookahead >= WANT_MIN_MATCH && s->strstart > 0) {
```
Do we actually need to change the minimum lookahead from 3 to 4 bytes here, why?
This doesn't use the 32bit hashing, so I don't quite see what the reason is for this change..?

Or is it so we have enough buffer space for running compare256_rle_unaligned32? If so, what then about the 64bit version?

nmoinvaz replied:
It does not need to be updated. I think at the time I was simply copying some change in the other deflate sources. I have changed it back to use `STD_MIN_MATCH`.

Dead2 commented on test/test_compare256_rle.cc:
```diff
@@ -0,0 +1,63 @@
+
+#ifdef UNALIGNED_OK
+TEST_COMPARE256_RLE(unaligned_16, compare256_rle_unaligned_16, 1)
+#ifdef HAVE_BUILTIN_CTZ
```
ifdef indentation?

nmoinvaz replied:
For function declarations we have not previously used indenting at least not in `cpu_features.h` or in the benchmark source code.  I just copied it from those sources.

nmoinvaz commented:
### DEVELOP https://github.com/zlib-ng/zlib-ng/commit/85f9709137a88c546e5dd188f7cb196b80f84253
```
OS: Windows 10 10.0.22000 AMD64
CPU: Intel64 Family 6 Model 158 Stepping 10, GenuineIntel
Tool: ../zlib-ng/build-1/Release/minigzip.exe
Levels: 0-0
Runs: 70         Trim worst: 40

 Level   Comp   Comptime min/avg/max/stddev  Decomptime min/avg/max/stddev  Compressed size
 0    100.008%      0.166/0.171/0.174/0.002        0.165/0.171/0.174/0.003      211,973,953
 R     60.465%      2.011/2.033/2.053/0.011        0.875/0.893/0.904/0.008      128,159,347

 avg1  80.236%                        1.102                          0.532
 avg2 160.472%                        2.204                          1.064
 tot                                 66.108                         31.930      340,133,300
```
### PR https://github.com/zlib-ng/zlib-ng/commit/ec80b4464f9ad73aa5c6ede9aed3a6ec07d4cae9
```
OS: Windows 10 10.0.22000 AMD64
CPU: Intel64 Family 6 Model 158 Stepping 10, GenuineIntel
Tool: ../zlib-ng/build-2/Release/minigzip.exe
Levels: 0-0
Runs: 70         Trim worst: 40

 Level   Comp   Comptime min/avg/max/stddev  Decomptime min/avg/max/stddev  Compressed size
 0    100.008%      0.158/0.165/0.168/0.002        0.163/0.170/0.175/0.004      211,973,953
 R     60.465%      1.981/2.005/2.027/0.015        0.841/0.868/0.879/0.009      128,159,347

 avg1  80.236%                        1.085                          0.519
 avg2 160.472%                        2.169                          1.038
 tot                                 65.079                         31.130      340,133,300
```
Results are not that great this time around. Not sure if this PR is worth it. It would be good to get other benchmarks from you guys to see.

nmoinvaz commented:
Rebased.

Dead2 commented:
I just had a renewed look at this, and considering the earlier unaligned debate, would it perhaps be better if we removed the unaligned_16 variant of this but kept the 32 and 64 variants?
That means the _c version becomes the fallback more often, but I would think the amount of systems that don't support the required builtins is already quite low, and thus there might be even less incentive to have the 16bit version.

There is also a comment referring to strend where strend is removed in deflate_rle.c.

If there is a rebase, should perhaps the first and the last commits be squashed together? At least from my quick look it looks like only new code is moved by the last commit, so it becomes a bit of unnecessary churn. Having the test in a separate commit is great though 👍

nmoinvaz committed:
Added support for unaligned/intrinsic to deflate_rle.

nmoinvaz committed:
Added benchmark for compare256_rle family of functions.

nmoinvaz commented:
> I just had a renewed look at this, and considering the earlier unaligned debate, would it perhaps be better if we removed the unaligned_16 variant of this but kept the 32 and 64 variants?

I think there is still some value because of the wide variety of systems that are being compiled against. It matches with `compare256_unaligned_16`. It shouldn't add too many bytes to code size either. It still ends up being faster than the C version (see [here](https://github.com/zlib-ng/zlib-ng/pull/810#issuecomment-1071954363)). However, if you want me to remove it I can.

> If there is a rebase, should perhaps the first and the last commits be squashed together?

Fixed.

phprus commented on deflate_rle.c:
```diff
@@ -36,19 +35,11 @@ Z_INTERNAL block_state deflate_rle(deflate_state *s, int flush) {
-            }
-            Assert(scan <= s->window + s->window_size - 1, "wild scan");
+            if (scan[0] == scan[1] && scan[1] == scan[2])
+                match_len = functable.compare256_rle(scan, scan+3)+2;
```
Is that the single point of calling ``functable.compare256_rle``?

If yes, then maybe make the functions ``inline`` and instead of calling through a pointer in the ``functable``, use direct name substitution (maybe by macros) at this line?

nmoinvaz replied:
Good point. `functable.compare256` is only exposed in `functable` because of its use in `deflate_quick`.

phprus replied:
Not ``functable.compare256`` because ``functable.compare256`` depends on CPU features:

https://github.com/zlib-ng/zlib-ng/blob/c72cd309cafae5d7e5eed42b9c4f36a53414362d/functable.c#L312-L323

It's about ``functable.compare256_rle``, because it depends only on macro values.

nmoinvaz replied:
I mentioned `compare256` because I remembered why that one was in `functable` and `compare256_rle` should not be. I also suppose you are right, in that it is also used there because of feature flags, not only because of usage of `deflate_quick`.

nmoinvaz replied:
Anyways, I have made some changes if you could review.

nmoinvaz commented:
Updated `compare256_rle` benchmarks:
```
Run on (20 X 2117.83 MHz CPU s)
CPU Caches:
  L1 Data 48 KiB (x10)
  L1 Instruction 32 KiB (x10)
  L2 Unified 1280 KiB (x10)
  L3 Unified 25600 KiB (x1)
--------------------------------------------------------------------------
Benchmark                                Time             CPU   Iterations
--------------------------------------------------------------------------
compare256_rle/c/1                   0.766 ns        0.125 ns   1000000000
compare256_rle/c/8                    1.81 ns        0.406 ns   1000000000
compare256_rle/c/64                   10.8 ns         1.15 ns    749294933
compare256_rle/c/256                  71.6 ns         5.78 ns    100000000
compare256_rle/unaligned_16/1        0.684 ns        0.094 ns   1000000000
compare256_rle/unaligned_16/8         1.40 ns        0.141 ns   1000000000
compare256_rle/unaligned_16/64        5.69 ns         1.03 ns    746666667
compare256_rle/unaligned_16/256       24.6 ns         6.56 ns    100000000
compare256_rle/unaligned_32/1         1.80 ns        0.297 ns   1000000000
compare256_rle/unaligned_32/8         2.93 ns        0.484 ns   1000000000
compare256_rle/unaligned_32/64        6.50 ns         1.12 ns   1000000000
compare256_rle/unaligned_32/256       20.7 ns         3.83 ns    167253334
compare256_rle/unaligned_64/1         2.02 ns        0.531 ns   1000000000
compare256_rle/unaligned_64/8         2.40 ns        0.750 ns   1000000000
compare256_rle/unaligned_64/64        5.86 ns         1.51 ns    746666667
compare256_rle/unaligned_64/256       10.1 ns         2.43 ns    373333333
```
Test 2
```
Run on (20 X 2122.12 MHz CPU s)
CPU Caches:
  L1 Data 48 KiB (x10)
  L1 Instruction 32 KiB (x10)
  L2 Unified 1280 KiB (x10)
  L3 Unified 25600 KiB (x1)
--------------------------------------------------------------------------
Benchmark                                Time             CPU   Iterations
--------------------------------------------------------------------------
compare256_rle/c/1                   0.700 ns        0.203 ns   1000000000
compare256_rle/c/8                    1.57 ns        0.531 ns   1000000000
compare256_rle/c/64                   9.70 ns         2.58 ns    218156522
compare256_rle/c/256                  65.4 ns         17.7 ns     34461538
compare256_rle/unaligned_16/1        0.667 ns        0.266 ns   1000000000
compare256_rle/unaligned_16/8         1.20 ns        0.266 ns   1000000000
compare256_rle/unaligned_16/64        5.10 ns         1.33 ns    494344828
compare256_rle/unaligned_16/256       23.3 ns         6.09 ns    100000000
compare256_rle/unaligned_32/1         1.66 ns        0.453 ns   1000000000
compare256_rle/unaligned_32/8         2.82 ns         1.08 ns   1000000000
compare256_rle/unaligned_32/64        5.90 ns         2.79 ns    280000000
compare256_rle/unaligned_32/256       20.1 ns         7.11 ns    112000000
compare256_rle/unaligned_64/1         1.90 ns        0.802 ns    896000000
compare256_rle/unaligned_64/8         2.19 ns         1.09 ns   1000000000
compare256_rle/unaligned_64/64        5.35 ns         2.65 ns    194782609
compare256_rle/unaligned_64/256       9.28 ns         5.47 ns    100000000
```

Dead2 commented:
@nmoinvaz Please rebase, then I'll merge this.
Tested as-is, and it does not trigger any Coverity warnings.

Dead2 reviewed (approved):
LGTM

Dead2 commented:
i9900K, cores locked to 3200Mhz, deep idle-states disabled, cpuset used to isolate cores from rest of system and functions aligned to 64b cachelines.

```
------------------------------------------------------------------------------------
Benchmark                                          Time             CPU   Iterations
------------------------------------------------------------------------------------

## Run 1
compare256_rle/c/1                              1.72 ns         1.72 ns    407783789
compare256_rle/c/8                              3.19 ns         3.19 ns    219427308
compare256_rle/c/64                             14.9 ns         14.9 ns     47035173
compare256_rle/c/256                            56.0 ns         56.0 ns     12508321
compare256_rle/unaligned_16/1                   2.01 ns         2.01 ns    348956731
compare256_rle/unaligned_16/8                   3.20 ns         3.19 ns    219269277
compare256_rle/unaligned_16/64                  12.8 ns         12.8 ns     54770855
compare256_rle/unaligned_16/256                 45.4 ns         45.4 ns     15437782
compare256_rle/unaligned_32/1                   1.99 ns         1.99 ns    352421855
compare256_rle/unaligned_32/8                   3.19 ns         3.19 ns    219766863
compare256_rle/unaligned_32/64                  11.4 ns         11.4 ns     61239149
compare256_rle/unaligned_32/256                 39.6 ns         39.5 ns     17702367
compare256_rle/unaligned_64/1                   2.01 ns         2.00 ns    349275656
compare256_rle/unaligned_64/8                   2.60 ns         2.60 ns    269232313
compare256_rle/unaligned_64/64                  6.72 ns         6.71 ns    104367968
compare256_rle/unaligned_64/256                 21.9 ns         21.8 ns     32043958

## Run 2
compare256_rle/c/1                              1.72 ns         1.72 ns    407744955
compare256_rle/c/8                              3.20 ns         3.19 ns    219385588
compare256_rle/c/64                             14.9 ns         14.9 ns     47043212
compare256_rle/c/256                            56.1 ns         56.0 ns     12503459
compare256_rle/unaligned_16/1                   2.01 ns         2.01 ns    348904805
compare256_rle/unaligned_16/8                   3.20 ns         3.19 ns    219222506
compare256_rle/unaligned_16/64                  12.8 ns         12.8 ns     54752555
compare256_rle/unaligned_16/256                 45.4 ns         45.4 ns     15433089
compare256_rle/unaligned_32/1                   1.99 ns         1.99 ns    352372028
compare256_rle/unaligned_32/8                   3.19 ns         3.19 ns    219683987
compare256_rle/unaligned_32/64                  11.5 ns         11.4 ns     61220688
compare256_rle/unaligned_32/256                 39.6 ns         39.6 ns     17696086
compare256_rle/unaligned_64/1                   2.01 ns         2.00 ns    349229976
compare256_rle/unaligned_64/8                   2.60 ns         2.60 ns    269202148
compare256_rle/unaligned_64/64                  6.72 ns         6.71 ns    104340182
compare256_rle/unaligned_64/256                 21.9 ns         21.9 ns     32031847
```

Looks like a clear win

nmoinvaz commented:
I wonder if it can be optimized further for the /1 cases.

nmoinvaz commented:
Instead of doing the byte shifts I've thought about doing two memory loads.

nmoinvaz commented:
Rebased

nmoinvaz commented:
> Instead of doing the byte shifts I've thought about doing two memory loads.

I tried this and it is much slower.

nmoinvaz commented:
I'm gtg on this PR.
